docs/puzzles/day3.md

```diff
@@ -0,0 +1,295 @@
+We are therefore computing the number of `1`'s.
+
+For each bit position, we then compare the total count of `1`'s to the total number of lines to compute `gammaRateBits`:
+
+```scala
+val total = bitLines.size
+
+val gammaRateBits: BitLine =
+  for (sumOfOneBits <- sumsOfOneBits)
+    yield (if (sumOfOneBits * 2 > total) 1 else 0)
+val gammaRate = bitLineToInt(gammaRateBits)
+
+def bitLineToInt(bitLine: BitLine): Int =
+  Integer.parseInt(bitLine.mkString, 2)
+```
+
+We use `sumOfOneBits * 2 > total` instead of `sumOfOneBits > total / 2` (which may be more intuitive) in order not to worry about the case where `total` is odd.
+`bitLineToInt` is nothing else than putting the bits back together as a string, as parsing it as an integer in base 2.
+
+We use a similar solution for `epsilonRate`.
+
+### Final code for part 1
+
+```scala
+def part1(input: String): Int =
+  val bitLines: List[BitLine] = input.linesIterator.map(parseBitLine).toList
+
+  val sumsOfOneBits: IndexedSeq[Int] = bitLines.reduceLeft((prevSum, line) =>
+    for ((prevBitSum, lineBit) <- prevSum.zip(line))
+      yield prevBitSum + lineBit
+  )
+  val total = bitLines.size // this will walk the list a second time, but that's OK
+
+  val gammaRateBits: BitLine =
+    for (sumOfOneBits <- sumsOfOneBits)
+      yield (if (sumOfOneBits * 2 > total) 1 else 0)
+  val gammaRate = bitLineToInt(gammaRateBits)
+
+  val epsilonRateBits: BitLine =
+    for (sumOfOneBits <- sumsOfOneBits)
+      yield (if (sumOfOneBits * 2 < total) 1 else 0)
+  val epsilonRate = bitLineToInt(epsilonRateBits)
+
+  gammaRate * epsilonRate
+
+type BitLine = IndexedSeq[Int]
+
+def parseBitLine(line: String): BitLine =
+  line.map(c => c - '0') // 1 or 0
+
+def bitLineToInt(bitLine: BitLine): Int =
+  Integer.parseInt(bitLine.mkString, 2)
+```
+
+<Solver puzzle="day3-part1"/>
+
+## First solution of Part 2
+
+Part 2 of this puzzle starts to get somewhat trickier.
+First, we solve it as naively as possible, by directly following the rules.
+
+```scala
+def part2(input: String): Int =
+  val bitLines = input.linesIterator.map(parseBitLine).toList
+
+  // bit criteria: keep the '1's if there is more '1's than '0's or if there is a tie
+  val oxygenGeneratorRatingLine: BitLine = recursiveFilter(bitLines, 0,
+      (totalOnes, total) => if (totalOnes * 2 >= total) 1 else 0)
+  val oxygenGeneratorRating = bitLineToInt(oxygenGeneratorRatingLine)
+
+  // bit criteria: keep the '0's if there is more '0's than '1's or if there is a tie
+  // equivalent to: keep the '1's if there is strictly less '1's than '0's
+  val co2ScrubberRatingLine: BitLine = recursiveFilter(bitLines, 0,
+      (totalOnes, total) => if (totalOnes * 2 < total) 1 else 0)
+  val co2ScrubberRating = bitLineToInt(co2ScrubberRatingLine)
+
+  oxygenGeneratorRating * co2ScrubberRating
+
+@scala.annotation.tailrec
+def recursiveFilter(bitLines: List[BitLine], bitPosition: Int,
+    bitCriteria: (Int, Int) => Int): BitLine =
```

Comment:
Perhaps I should already use `keepMostCommon: Boolean` in this first version, rather than the function criteria, since it doesn't generalize well in the improved version?